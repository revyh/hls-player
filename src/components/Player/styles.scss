@import 'settings.scss';

.base {
  display: block;
  width: 100%;
  height: 100%;
  position: relative;
  background-color: $base-bg-color;
}

.loader-elem {
  display: block;
  position: absolute;
  width: 100px;
  height: 100px;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: 2;
}

.loader-icon-elem {
  display: block;
  position: absolute;
  width: 50px;
  height: 50px;
  top: 50%;
  left: 30px;
  transform: translateY(-50%);
  fill: #fff;
}

.loader-btn-elem {
  display: block;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  cursor: pointer;
}

.title-elem {
  position: absolute;
  top: 0;
  display: block;
  height: 190px;
  width: 100%;
  padding: 66px 50px 0;
  box-sizing: border-box;
  background-image:
    linear-gradient(
      to bottom,
      rgba($base-bg-color, 1) 0%,
      rgba($base-bg-color, 0.7) 20%,
      rgba($base-bg-color, 0) 100%
    );
  z-index: 1;
}

.title-elem__hidden {
  composes: title-elem;
  display: none;
}

.title-text-elem {
  display: block;
  font-size: 30px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.panel-elem {
  position: absolute;
  bottom: 0;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  align-items: stretch;
  height: 190px;
  width: 100%;
  padding: 0 30px 66px;
  box-sizing: border-box;
  background-image:
    linear-gradient(
      to top,
      rgba($base-bg-color, 1) 0%,
      rgba($base-bg-color, 0.7) 20%,
      rgba($base-bg-color, 0) 100%
    );
  z-index: 1;
}

.panel-elem__hidden {
  composes: panel-elem;
  display: none;
}

.video-elem {
  display: block;
  position: absolute;
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.video-elem__hidden {
  composes: video-elem;
  opacity: 0;
}
