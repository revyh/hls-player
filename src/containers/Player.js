import {connect} from 'react-redux';
import {createSelector} from 'reselect';
import {
  actions as playbackActions,
  selectors as playbackSelectors,
} from 'features/playback';
import {
  actions as mediaActions,
  selectors as mediaSelectors,
} from 'features/media';
import * as selectors from 'features/selectors';
import Presentational from 'components/Player';

const getProps = createSelector(
  createSelector(selectors.playback, playbackSelectors.playback),
  createSelector(selectors.media, mediaSelectors.media),
  (playback, media) => ({...playback, ...media}),
);

function mapStateToProps(state, props) {
  return {
    ...getProps(state),
    ...props,
  };
}

const mapDispatchToProps = {
  onPauseChange: playbackActions.setPause,
  onMuteChange: playbackActions.setMute,
  onVolumeChange: playbackActions.setVolume,
  onTimeChange: playbackActions.setTime,
  onFullscreenChange: playbackActions.setFullscreenMode,
  onLoad: mediaActions.setMedia,
};

export default connect(mapStateToProps, mapDispatchToProps)(Presentational);
