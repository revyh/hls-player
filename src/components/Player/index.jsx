import React from 'react';
import screenfull from 'screenfull';
import styles from './styles.scss';
import Panel from 'components/Panel';
import Video from 'components/Video';
import Spinner from 'components/Spinner';
import Icon from 'components/Icon';
import getClassName from 'utils/getClassName';
import workInAnimationFrame from 'utils/workInAnimationFrame';

const CONTROLS_VISIBILITY_TIMEOUT = 5000;

export default class Player extends React.Component {
  state = {
    isControlsVisible: false,
  };

  componentDidMount() {
    this.updateClient(this.props.Client);
    document.addEventListener('mousemove', this.checkControlsVisibility);
    document.addEventListener(
      screenfull.raw.fullscreenchange,
      this.handleFullscreenEscape,
    );
  }

  componentDidUpdate(prevProps) {
    this.updateClient(prevProps.Client);
    this.syncFullscreen();
  }

  componentWillUnmount() {
    if (this.client)
      this.client.destroy();

    document.removeEventListener('mousemove', this.checkControlsVisibility);
    document.removeEventListener(
      screenfull.raw.fullscreenchange,
      this.handleFullscreenEscape,
    );
  }

  get isLoaded() {
    return this.client && this.client.isLoaded;
  }

  getRef = ref => {
    this.containerElement = ref;
    this.syncFullscreen();
  };

  getVideoRef = ref => {
    this.videoElement = ref;
    this.updateClient(this.props.Client);
  };

  handleFullscreenEscape = () => {
    if (!screenfull.isFullscreen)
      this.props.onFullscreenChange(false);
  };

  handleLoad = () => {
    const {mediaUrl, onLoad} = this.props;
    onLoad && onLoad(mediaUrl);
  };

  checkControlsVisibility = workInAnimationFrame(() => {
    if (this.controlsVisibilityTimeout)
      clearTimeout(this.controlsVisibilityTimeout);

    if (!this.state.isControlsVisible)
      this.setState({...this.state, isControlsVisible: true});

    this.controlsVisibilityTimeout = setTimeout(
      () => { this.setState({...this.state, isControlsVisible: false}); },
      CONTROLS_VISIBILITY_TIMEOUT,
    );
  });

  syncFullscreen() {
    const {isFullscreenEntered} = this.props;
    const container = this.containerElement;

    if (!container)
      return;

    if (isFullscreenEntered && screenfull.element !== container)
      screenfull.request(container);
    else if (!isFullscreenEntered && screenfull.element === container)
      screenfull.exit();
  }

  updateClient(prevClient) {
    const {Client, url} = this.props;
    const videoElement = this.videoElement;

    if (this.client)
      if (
        this.client.url === url
        && this.client.videoElement === videoElement
        && prevClient === Client
      )
        return;
      else {
        this.client.destroy();
        this.client = null;
      }

    if (!url || !videoElement || !Client)
      return;

    this.client = new Client();
    this.client.start(url, this.videoElement);
  }

  renderLoader() {
    return this.isLoaded
      ? null
      : (
        <div className={styles.loaderElem}>
          <Spinner active={!!this.props.url} />
          <Icon className={styles.loaderIconElem} symbol="play" />
          <button
            className={styles.loaderBtnElem}
            onClick={!this.props.url && this.handleLoad}
          />
        </div>
      );
  }

  renderTitle() {
    const titleClassName = this.isLoaded && this.state.isControlsVisible
      ? styles.titleElem
      : styles.titleElemHidden;

    return (
      <div className={titleClassName}>
        <span className={styles.titleTextElem}>{this.props.title}</span>
      </div>
    );
  }

  renderVideo() {
    const {isPaused, isMuted, volume, onTimeChange} = this.props;

    const videoClassName = this.isLoaded
      ? styles.videoElem
      : styles.videoElemHidden;

    return (
      <Video
        className={videoClassName}
        isPaused={isPaused}
        isMuted={isMuted}
        volume={volume}
        onTimeChange={onTimeChange}
        videoRef={this.getVideoRef}
      />
    );
  }

  renderPanel() {
    const {
      isPaused,
      isMuted,
      volume,
      isFullscreenEntered,
      time,
      onPauseChange,
      onMuteChange,
      onVolumeChange,
      onFullscreenChange,
    } = this.props;

    const isVisible = this.state.isControlsVisible || isFullscreenEntered;
    const panelClassName = this.isLoaded && isVisible
      ? styles.panelElem
      : styles.panelElemHidden;

    return (
      <div className={panelClassName}>
        <Panel
          isPaused={isPaused}
          isMuted={isMuted}
          isFullscreenEntered={isFullscreenEntered}
          volume={volume}
          time={time}
          onPauseChange={onPauseChange}
          onMuteChange={onMuteChange}
          onVolumeChange={onVolumeChange}
          onFullscreenChange={onFullscreenChange}
        />
      </div>
    );
  }

  render() {
    return (
      <div
        className={getClassName(styles, this.props, styles.base)}
        ref={this.getRef}
      >
        {this.renderLoader()}
        {this.renderTitle()}
        {this.renderVideo()}
        {this.renderPanel()}
      </div>
    );
  }
}
