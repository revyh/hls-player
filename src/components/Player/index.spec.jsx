import React from 'react';
import {shallow, mount} from 'enzyme';
import Player from '.';

// eslint-disable-next-line no-var
var mockScreenfull;

jest.mock('components/Icon');
jest.mock('utils/workInAnimationFrame', () => arg => arg);
jest.mock('screenfull', () => {
  mockScreenfull = mockScreenfull || {
    element: null,
    request: jest.fn(),
    exit: jest.fn(),
    raw: {fullscreenchange: 'fullscreenchange'},
  };
  return mockScreenfull;
});

jest.useFakeTimers();

beforeEach(() => {
  mockScreenfull.element = null;
  mockScreenfull.request.mockClear();
  mockScreenfull.exit.mockClear();
});

const MockClient = jest.fn(function MockClient() {
  this.startPromise = Promise.resolve();

  this.start = jest.fn((startUrl, videoElement) => {
    this.url = startUrl;
    this.videoElement = videoElement;

    return this.startPromise.then(() => {
      this.isLoaded = true;
    });
  });

  this.destroy = jest.fn();
});

function dispatchMouseMove() {
  const mouseMoveEvent = document.createEvent('MouseEvents');
  mouseMoveEvent.initEvent('mousemove', true, true);
  document.dispatchEvent(mouseMoveEvent);
}

function expectControlsVisible(player) {
  expect(player.find('.titleElem').exists()).toBeTruthy();
  expect(player.find('.panelElem').exists()).toBeTruthy();
}

function expectControlsHidden(player) {
  expect(player.find('.titleElemHidden').exists()).toBeTruthy();
  expect(player.find('.panelElemHidden').exists()).toBeTruthy();
}

function renderStub(props) {
  return shallow(
    <Player
      Client={MockClient}
      url="some-url"
      title="some title"
      isPaused
      isMuted
      isFullscreenEntered
      volume={10}
      time={3600}
      onPauseChange={jest.fn()}
      onMuteChange={jest.fn()}
      onVolumeChange={jest.fn()}
      onFullscreenChange={jest.fn()}
      {...props}
    />,
  );
}

it('should render', () => {
  expect(renderStub({url: null})).toMatchSnapshot();
});

it('should render in loading state', () => {
  expect(renderStub()).toMatchSnapshot();
});

it('should render in loaded state', () => {
  const player = renderStub();
  player.instance().getVideoRef('videoElement');
  MockClient.mock.instances[0].isLoaded = true;
  player.instance().forceUpdate();

  expect(player.shallow()).toMatchSnapshot();
});

it('should render with visible controls in loaded state', () => {
  const player = renderStub();
  player.instance().getVideoRef('videoElement');
  MockClient.mock.instances[0].isLoaded = true;
  player.setState({isControlsVisible: true});

  expect(player.shallow()).toMatchSnapshot();
});

it('should init new client', () => {
  mount(<Player Client={MockClient} url="some-url" />);
  const clientMock = MockClient.mock.instances[0];

  expect(clientMock.start).toHaveBeenCalledTimes(1);
  expect(clientMock.start.mock.calls[0][0]).toBe('some-url');
});

it('should destroy previous client', () => {
  const player = mount(<Player Client={MockClient} url="some-url" />);
  const clientMock = MockClient.mock.instances[0];

  player.setProps({url: 'another-url'});

  expect(clientMock.destroy).toHaveBeenCalled();
});

it('should reinit client only when needed', () => {
  const player = mount(<Player Client={MockClient} url="some-url" />);
  const clientMock = MockClient.mock.instances[0];

  player.update();

  expect(clientMock.start).toHaveBeenCalledTimes(1);
  expect(clientMock.destroy).not.toHaveBeenCalled();
});

it('should show hidden controls on \'mousemove\' event', () => {
  const player = mount(<Player Client={MockClient} url="some-url" />);
  MockClient.mock.instances[0].isLoaded = true;

  expectControlsHidden(player);

  dispatchMouseMove();
  expectControlsVisible(player);
});

it('should hide controls on timeout after \'mousemove\' event', () => {
  const player = mount(<Player Client={MockClient} url="some-url" />);
  MockClient.mock.instances[0].isLoaded = true;

  dispatchMouseMove();
  expectControlsVisible(player);

  jest.runAllTimers();
  expectControlsHidden(player);
});

it('should rerun timer on each \'mousemove\' event', () => {
  const player = mount(<Player Client={MockClient} url="some-url" />);
  MockClient.mock.instances[0].isLoaded = true;

  dispatchMouseMove();
  expectControlsVisible(player);

  jest.runTimersToTime(4000);
  dispatchMouseMove();
  jest.runTimersToTime(4000);

  expectControlsVisible(player);
});

it('should enter fullscreen', () => {
  const player = mount(<Player />);
  const containerElement = player.instance().containerElement;

  player.setProps({isFullscreenEntered: true});

  expect(mockScreenfull.request).toHaveBeenCalledWith(containerElement);
});

it('should leave fullscreen', () => {
  const player = mount(<Player isFullscreenEntered />);
  mockScreenfull.element = player.instance().containerElement;

  player.setProps({isFullscreenEntered: false});

  expect(mockScreenfull.exit).toHaveBeenCalled();
});

it('should show controls in fullscreen mode', () => {
  const player = mount(<Player Client={MockClient} url="some-url" />);
  MockClient.mock.instances[0].isLoaded = true;

  expect(player.find('.titleElemHidden').exists()).toBeTruthy();
  expect(player.find('.panelElemHidden').exists()).toBeTruthy();

  player.setProps({isFullscreenEntered: true});

  expect(player.find('.titleElemHidden').exists()).toBeTruthy();
  expect(player.find('.panelElem').exists()).toBeTruthy();
});

// skip because `fullscreenchange` event doesn't implemented in jsdom yet
it.skip('should trigger \'onFullscreenChange\' event handler on \'esc\' key', () => {
  const mockOnFullscreenChange = jest.fn();
  mount(
    <Player
      isFullscreenEntered
      onFullscreenChange={mockOnFullscreenChange}
    />,
  );

  const fullscreenChangeEvent = document.createEvent('ViewEvents');
  fullscreenChangeEvent.initEvent('fullscreenchange', true, true);
  document.dispatchEvent(fullscreenChangeEvent);

  expect(mockOnFullscreenChange).toHaveBeenCalledWith(false);
});

it('should trigger \'onLoad\' event handler', () => {
  const mockOnLoad = jest.fn();
  const player = mount(
    <Player
      mediaUrl="some-media-url"
      onLoad={mockOnLoad}
    />,
  );

  player.find('.loaderBtnElem').simulate('click');

  expect(mockOnLoad).toHaveBeenCalledWith('some-media-url');
});
