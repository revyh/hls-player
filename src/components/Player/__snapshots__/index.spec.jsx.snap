// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="base"
>
  <div
    className="loaderElem"
  >
    <Spinner
      active={false}
    />
    <Icon
      className="loaderIconElem"
      symbol="play"
    />
    <button
      className="loaderBtnElem"
      onClick={[Function]}
    />
  </div>
  <div
    className="titleElemHidden"
  >
    <span
      className="titleTextElem"
    >
      some title
    </span>
  </div>
  <Video
    className="videoElemHidden"
    isMuted={true}
    isPaused={true}
    videoRef={[Function]}
    volume={10}
  />
  <div
    className="panelElemHidden"
  >
    <Panel
      isFullscreenEntered={true}
      isMuted={true}
      isPaused={true}
      onFullscreenChange={[Function]}
      onMuteChange={[Function]}
      onPauseChange={[Function]}
      onVolumeChange={[Function]}
      time={3600}
      volume={10}
    />
  </div>
</div>
`;

exports[`should render in loaded state 1`] = `
<div
  className="base"
>
  <div
    className="titleElemHidden"
  >
    <span
      className="titleTextElem"
    >
      some title
    </span>
  </div>
  <Video
    className="videoElem"
    isMuted={true}
    isPaused={true}
    videoRef={[Function]}
    volume={10}
  />
  <div
    className="panelElem"
  >
    <Panel
      isFullscreenEntered={true}
      isMuted={true}
      isPaused={true}
      onFullscreenChange={[Function]}
      onMuteChange={[Function]}
      onPauseChange={[Function]}
      onVolumeChange={[Function]}
      time={3600}
      volume={10}
    />
  </div>
</div>
`;

exports[`should render in loading state 1`] = `
<div
  className="base"
>
  <div
    className="loaderElem"
  >
    <Spinner
      active={true}
    />
    <Icon
      className="loaderIconElem"
      symbol="play"
    />
    <button
      className="loaderBtnElem"
      onClick={false}
    />
  </div>
  <div
    className="titleElemHidden"
  >
    <span
      className="titleTextElem"
    >
      some title
    </span>
  </div>
  <Video
    className="videoElemHidden"
    isMuted={true}
    isPaused={true}
    videoRef={[Function]}
    volume={10}
  />
  <div
    className="panelElemHidden"
  >
    <Panel
      isFullscreenEntered={true}
      isMuted={true}
      isPaused={true}
      onFullscreenChange={[Function]}
      onMuteChange={[Function]}
      onPauseChange={[Function]}
      onVolumeChange={[Function]}
      time={3600}
      volume={10}
    />
  </div>
</div>
`;

exports[`should render with visible controls in loaded state 1`] = `
<div
  className="base"
>
  <div
    className="titleElem"
  >
    <span
      className="titleTextElem"
    >
      some title
    </span>
  </div>
  <Video
    className="videoElem"
    isMuted={true}
    isPaused={true}
    videoRef={[Function]}
    volume={10}
  />
  <div
    className="panelElem"
  >
    <Panel
      isFullscreenEntered={true}
      isMuted={true}
      isPaused={true}
      onFullscreenChange={[Function]}
      onMuteChange={[Function]}
      onPauseChange={[Function]}
      onVolumeChange={[Function]}
      time={3600}
      volume={10}
    />
  </div>
</div>
`;
